TeleCall uses 4 centers around the globe to process customer order forms. They audit a certain %  of the customer order forms.
Any error in order form renders it defective and has to be reworked before processing.  
The manager wants to check whether the defective %  varies by centre.


Chi-square test of independence
Defining hypothesis test:
H0: all center defective % are same, does not vary by center
Ha: at least one defective %  varies by centre

In [54]:
import pandas as pd
import numpy as np
import scipy
from scipy import stats

In [3]:
data_custOrder = pd.read_csv('Costomer+OrderForm.csv')
data_custOrder.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Phillippines  300 non-null    object
 1   Indonesia     300 non-null    object
 2   Malta         300 non-null    object
 3   India         300 non-null    object
dtypes: object(4)
memory usage: 9.5+ KB


In [27]:
data_custOrder.head()

Phillippines   Indonesia       Malta       India
0   Error Free  Error Free   Defective  Error Free
1   Error Free  Error Free  Error Free   Defective
2   Error Free   Defective   Defective  Error Free
3   Error Free  Error Free  Error Free  Error Free
4   Error Free  Error Free   Defective  Error Free

In [48]:
print(data_custOrder.Phillippines.value_counts(),'\n',
        data_custOrder.Indonesia.value_counts(),'\n',
        data_custOrder.Malta.value_counts(),'\n',
        data_custOrder.India.value_counts())

Error Free    271
Defective      29
Name: Phillippines, dtype: int64 
 Error Free    267
Defective      33
Name: Indonesia, dtype: int64 
 Error Free    269
Defective      31
Name: Malta, dtype: int64 
 Error Free    280
Defective      20
Name: India, dtype: int64


In [52]:
#contingency table/ observed val

obs_val = np.array([[271, 267, 269, 280],
                    [29, 33, 31, 20]])
obs_val

array([[271, 267, 269, 280],
       [ 29,  33,  31,  20]])

In [63]:
#[chi2 test val, p val, df, expected val]
test_res = stats.chi2_contingency(obs_val)

print('chi2 test values:      ', test_res[0], 
      '\nP value:               ', test_res[1],
      '\ndegree of freedom:     ', test_res[2],
      '\nexpected values table: ', test_res[3])

chi2 test values:       3.858960685820355 
P value:                0.2771020991233135 
degree of freedom:      3 
expected values table:  [[271.75 271.75 271.75 271.75]
 [ 28.25  28.25  28.25  28.25]]


p value(0.277) > alfa(0.05); thus accept the null hypothesis.
H0: all center defective % are same, does not vary by center
